libc/string_catdup.c

```diff
@@ -0,0 +1,728 @@
+ *
+ * %LICENSE%
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include <ctype.h>
+#include <limits.h>
+#include <unity_fixture.h>
+#include "common.h"
+
+
+#define WORD_1          "Lorem "
+#define WORD_2          "ipsum"
+#define NON_ASCII_1     "\xe3\x83\x9e\xe3\x83\xaa\xe3\x82\xa2\xe3\x83\xbb"
+#define NON_ASCII_2     "\xe3\x82\xb9\xe3\x82\xaf\xe3\x82\xa6\xe3\x82\xa9\xe3\x83\x89\xe3\x83\x95\xe3\x82\xb9\xe3\x82\xab\xef\xbc\x9d\xe3\x82\xad\xe3\x83\xa5\xe3\x83\xaa\xe3\x83\xbc"
+#define SPECIALS_STRING "\n\t !\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~"
+#define SIZE_BUF        32
+#define SIZE_BIGGER_BUF 150
+
+
+static char huge_buf[PATH_MAX] = { 0 };
+
+
+TEST_GROUP(string_cat);
+TEST_GROUP(string_dup);
+
+
+TEST_SETUP(string_cat)
+{
+}
+
+
+TEST_TEAR_DOWN(string_cat)
+{
+}
+
+
+TEST(string_cat, strcat_basic)
+{
+	int i;
+	char buf[SIZE_BUF] = WORD_1;
+	char buf1[] = WORD_2;
+
+	TEST_ASSERT_EQUAL_PTR(buf, strcat(buf, buf1));
+	TEST_ASSERT_EQUAL_STRING(WORD_1 WORD_2, buf);
+
+	for (i = strlen(WORD_1 WORD_2); i < SIZE_BUF; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+
+TEST(string_cat, strcat_empty_dest)
+{
+	int i;
+	char buf[SIZE_BUF] = "";
+	char buf1[] = WORD_2;
+
+	TEST_ASSERT_EQUAL_PTR(buf, strcat(buf, buf1));
+	TEST_ASSERT_EQUAL_STRING(WORD_2, buf);
+
+	for (i = strlen(WORD_2); i < SIZE_BUF; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+
+TEST(string_cat, strcat_empty_src)
+{
+	int i;
+	char buf[SIZE_BUF] = WORD_1;
+	char buf1[] = "";
+
+	TEST_ASSERT_EQUAL_PTR(buf, strcat(buf, buf1));
+	TEST_ASSERT_EQUAL_STRING(WORD_1, buf);
+
+	for (i = strlen(WORD_1); i < SIZE_BUF; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+
+TEST(string_cat, strcat_empty)
+{
+	int i;
+	char buf[SIZE_BUF / 2] = "";
+	char buf1[] = "";
+
+	TEST_ASSERT_EQUAL_PTR(buf, strcat(buf, buf1));
+	TEST_ASSERT_EQUAL_STRING("", buf);
+
+	for (i = 0; i < SIZE_BUF / 2; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+
+TEST(string_cat, strcat_huge_string_dest)
+{
+	char buf[PATH_MAX] = "";
+	char buf1[SIZE_BUF] = "";
+
+	memcpy(buf, hugeTestStr, PATH_MAX - 1);
+
+	TEST_ASSERT_EQUAL_PTR(buf, strcat(buf, buf1));
+	TEST_ASSERT_EQUAL_STRING(huge_buf, buf);
+}
+
+
+TEST(string_cat, strcat_huge_string_src)
+{
+	char buf[PATH_MAX] = "";
+
+	TEST_ASSERT_EQUAL_PTR(buf, strcat(buf, huge_buf));
+	TEST_ASSERT_EQUAL_STRING(huge_buf, buf);
+}
+
+
+TEST(string_cat, strcat_huge_string_both)
+{
+	char buf[PATH_MAX] = "";
+	char buf1[PATH_MAX / 2] = "";
+
+	memcpy(buf, huge_buf, PATH_MAX / 2);
+	memcpy(buf1, &huge_buf[PATH_MAX / 2], PATH_MAX / 2 - 1);
+
+	TEST_ASSERT_EQUAL_PTR(buf, strcat(buf, buf1));
+	TEST_ASSERT_EQUAL_STRING(huge_buf, buf);
+}
+
+
+TEST(string_cat, strcat_specials_string)
+{
+	int i;
+	char buf[SIZE_BUF * 4] = SPECIALS_STRING;
+	char buf1[] = SPECIALS_STRING;
+
+	TEST_ASSERT_EQUAL_PTR(buf, strcat(buf, buf1));
+	TEST_ASSERT_EQUAL_STRING(SPECIALS_STRING SPECIALS_STRING, buf);
```

Comment:
nitpicking: instead of repeating result string twice in each test it might be better to explicitly:
`const char *expectedStr = SPECIALS_STRING SPECIALS_STRING;`

and then use it.